Guide on how to code an NFT smart contract

smudgil opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

This tutorial describes how to delpy an NFT smart contract. 

<!--- Describe your changes in detail -->

It walks the user step by step on this process. I will also submit a part 2, where the user can mint an NFT by interacting a smart contract and part 3,  where the user can view their smart contract in metamask!

<!--- This project accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+
+    {"jsonrpc": "2.0", "id": 0, "result": "0xde0b6b3a7640000"}
+
+**NOTE: **This result is in wei not eth. Wei is used as the smallest denomination of ether. The conversion from wei to eth is: 1 eth = 1⁰¹⁸ wei. So if we convert 0xde0b6b3a7640000 to decimal we get 1*1⁰¹⁸ which equals 1 eth.
```
- Instead of `1⁰¹⁸ wei` use `10<sup>18</sup> wei`
- Instead of `from wei to eth is: 1 eth...`
  use `from wei to ether is 1 ether...` or `from wei to ETH is 1 ETH...`

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+
+First, we’ll need to create a folder for our project. Navigate to your command line and type:
+
+    mkdir my-nft
+    cd my-nft
```
Wrap multi-line code snippets in triple-backticks. 

https://raw.githubusercontent.com/ethereum/ethereum-org-website/dev/src/content/developers/tutorials/run-light-node-geth/index.md
See some examples in there (optional is to place the programming language next to the opening set of triple-backticks as in that file.

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+
+Now that we’re inside our project folder, we’ll use npm init to initialize the project. If you don’t already have npm installed, follow [these instructions](https://docs.alchemyapi.io/alchemy/guides/alchemy-for-macs#1-install-nodejs-and-npm) (we’ll also need [Node.js](https://nodejs.org/en/download/), so download that too!).
+
+    npm init
```
Wrap in triple-backticks as well to present as a code block.

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+
+It doesn’t really matter how you answer the installation questions, here is how we did it for reference:
+
+    package name: (my-nft) 
+    version: (1.0.0) 
+    description: My first NFT!
+    entry point: (index.js) 
+    test command: 
+    git repository: 
+    keywords: 
+    author: 
+    license: (ISC) 
+    About to write to /Users/thesuperb1/Desktop/my-nft/package.json:
+
+    {
+      "name": "my-nft",
+      "version": "1.0.0",
+      "description": "My first NFT!",
+      "main": "index.js",
+      "scripts": {
+        "test": "echo \"Error: no test specified\" && exit 1"
+      },
+      "author": "",
+      "license": "ISC"
+    }
```
(Same as above here regarding code blocks)

wackerow replied:
..and throughout below

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+    mkdir contracts
+    mkdir scripts
+
+* contracts/ is where we’ll keep our NFT smart contract code
```
Can also use single backticks to wrap inline code snippets of file/path names

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+
+1. Navigate to the “contracts” folder and create a new file called MyNFT.sol
+
+1. Below is our NFT smart contract code, which based off of the [OpenZepplin](https://docs.openzeppelin.com/contracts/3.x/erc721) library’s ERC721 implementation. Copy and paste the contents below into your MyNFT.sol file.
```
`1.`  should be `2.`

Also, trying to standardize using hyphenated ERC references throughout site, ie. `ERC-721`... the mentions of ERC721 within code blocks below are fine as is

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+        }
+    }
+
+3. Because we are inheriting classes from the OpenZepplin contracts library, in your command line run npm install @openzeppelin/contracts to install the library into our folder.
```
Again, would add in some inline code snippet styling with backticks to help these commands stand out.

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+
+3. Because we are inheriting classes from the OpenZepplin contracts library, in your command line run npm install @openzeppelin/contracts to install the library into our folder.
+
+So, what does this code *do* exactly? Let’s break it down, line by line.
```
"line-by-line"

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+
+At the top of our smart contract, we import three [OpenZepplin](https://openzeppelin.com/) smart contract classes:
+
+* @openzeppelin/contracts/token/ERC721/ERC721.solcontains the implementation of the ERC721 standard, which our NFT smart contract will inherit. (To be a valid NFT, your smart contract must implement all the methods of the ERC721 standard.) To learn more about the inherited ERC721 functions, check out the interface definition [here](https://eips.ethereum.org/EIPS/eip-721).
```
"ERC-721" (and throughout) =)

wackerow replied:
And same w/ the inline styling throughout these sections

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+
+## Step 12: Install Ethers.js
+
+Ethers.js is a library that makes it easier to interact and make requests to Ethreum by wrapping [standard JSON-RPC methods](https://docs.alchemyapi.io/alchemy/documentation/alchemy-api-reference/json-rpc) with more user friendly methods.
```
JSON-RPC methods are hosted locally now, would link within:
[/developers/docs/apis/json-rpc/](https://ethereum.org/en/developers/docs/apis/json-rpc/)

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+
+Now that our contract is written and our configuration file is good to go, it’s time to write our contract deploy script.
+
+Navigate to the scripts/ folder and create a new file called deploy.js , adding the following contents to it:
```
Navigate to the `scripts/` folder and create a new file called `deploy.js`, adding the following contents to it:

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft.md:
```diff
@@ -0,0 +1,334 @@
+
+Here you’ll see a handful of JSON-RPC calls that Hardhat/Ethers made under the hood for us when we called the .deploy() function. Two important ones to call out here are [eth_sendRawTransaction](https://docs.alchemyapi.io/alchemy/documentation/alchemy-api-reference/json-rpc#eth_sendrawtransaction), which is the request to actually write our smart contract onto the Ropsten chain, and [eth_getTrans ...
+
+That’s all for Part 1 of this tutorial. In Part 2, we’ll actually interact with our smart contract by minting an NFT, and in Part 3 we’ll show you how to view your NFT in your Ethereum wallet!
```
Would add a link here to go to the next part

Don't worry about it being in a separate PR, they will end up merged together. Make sure to update the files in the other two PRs to `part-2-name/index.md` instead of `part-2-name.md` (for example), and in this PR, add a link to the bottom to `/developers/tutorials/part-2-name/`

wackerow reviewed (changes requested):
@smudgil This looks fantastic! What a great start. I left a bunch of comments throughout, many semantic, just trying to get this ready for production. Do me a favor and use these comments as a guide to touch up the other two PRs as well (#2653, #2654), and then I'll be happy to take another run through them all. 👍 Awesome stuff

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md:
```diff
@@ -0,0 +1,341 @@
+1. Navigate to the “contracts” folder and create a new file called MyNFT.sol
+
+2. Below is our NFT smart contract code, which based off of the [OpenZepplin](https://docs.openzeppelin.com/contracts/3.x/erc721) library’s ERC-721 implementation. Copy and paste the contents below into your MyNFT.sol file.
+
+    //Contract based on [https://docs.openzeppelin.com/contracts/3.x/erc721](https://docs.openzeppelin.com/contracts/3.x/erc721)
+    // SPDX-License-Identifier: MIT
+    pragma solidity ^0.7.3;
+
+    import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+    import "@openzeppelin/contracts/utils/Counters.sol";
+    import "@openzeppelin/contracts/access/Ownable.sol";
+    
+
+    contract MyNFT is ERC721, Ownable {
+        using Counters for Counters.Counter;
+        Counters.Counter private _tokenIds;
+
+        constructor() public ERC721("MyNFT", "NFT") {}
+
+        function mintNFT(address recipient, string memory tokenURI)
+            public onlyOwner
+            returns (uint256)
+        {
+            _tokenIds.increment();
+
+            uint256 newItemId = _tokenIds.current();
+            _mint(recipient, newItemId);
+            _setTokenURI(newItemId, tokenURI);
+
+            return newItemId;
+        }
+    }
+
```
So our markdown parser can sometimes get the indentation syntax correct, but it's safer to wrap these code blocks in triple-backticks. You can also add the language that way to help with syntax coloring.

**This is how this section renders now:**
![image](https://user-images.githubusercontent.com/54227730/115784032-733e6400-a372-11eb-97d5-ca50a3071a0e.png)
**If you wrap this block (and others) in triple-backticks (on their own lines) and trail the opening set with the language name, like this...**
![image](https://user-images.githubusercontent.com/54227730/115784411-eba52500-a372-11eb-8087-0649eff01c2e.png)
**...then it will render like this:**
![image](https://user-images.githubusercontent.com/54227730/115784464-fe1f5e80-a372-11eb-8e1f-2ad92a8b898e.png)

Lets make sure all the code blocks are wrapped this way to minimize issues and keep it consistent =) thanks!

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md:
```diff
@@ -0,0 +1,341 @@
+
+In our ERC-721 constructor, you’ll notice we pass 2 strings, “MyNFT” and “NFT.” The first variable is the smart contract’s name, and the second is its symbol. You can name each of these variables whatever you wish!
+
+Finally, we have our function mintNFT(address recipient, string memory tokenURI) that allows us to mint an NFT! You'll notice it this function takes in two variables:
```
I'll leave this to your judgement, but I'm always a fan of inline code being styled as such by wrapping in single backticks inline. 
![image](https://user-images.githubusercontent.com/54227730/115784929-9e758300-a373-11eb-9185-db2418f0daf4.png)

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md:
```diff
@@ -0,0 +1,341 @@
+
+Finally, we have our function mintNFT(address recipient, string memory tokenURI) that allows us to mint an NFT! You'll notice it this function takes in two variables:
+
+* address recipient specifies the address that will receive your freshly minted NFT
```
Same here for example, as this looks like it should be capitalized at the beginning until you realize that `address` is referring to a variable, which would've been easier to spot with code styling

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md:
```diff
@@ -0,0 +1,341 @@
+
+* string memory tokenURI is a string that should resolve to a JSON document that describes the NFT's metadata. An NFT's metadata is really what brings it to life, allowing it to have configurable properties, such as a name, description, image, and other attributes. In part 2 of this tutorial, we will describe how to configure this metadata.
+
+mintNFT calls some methods from the inherited ERC-721 library, and ultimately returns a number that represents the ID of the freshly minted NFT.
```
...and here. In particular, it's starting the sentence, so it should be capitalized, or styled as code to clarify.

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md:
````diff
@@ -0,0 +1,341 @@
+
+Navigate to the `scripts/` folder and create a new file called `deploy.js` , adding the following contents to it:
+
+```
````
Like I mentioned before, you could also add `js` immediately following the opening set of triple backticks and then the syntax will be highlighted as JavaScript.

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md:
```diff
@@ -0,0 +1,341 @@
+
+Now that our contract is written and our configuration file is good to go, it’s time to write our contract deploy script.
+
+Navigate to the `scripts/` folder and create a new file called `deploy.js` , adding the following contents to it:
```
remove extra space after `deploy.js`
```
Navigate to the `scripts/` folder and create a new file called `deploy.js`, adding the following contents to it:
```

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md:
```diff
@@ -0,0 +1,341 @@
+
+![View calls made “under the hood” with Alchemy’s Explorer Dashboard](https://cdn-images-1.medium.com/max/5200/0*9vAulgWSnJJ_Xf4u.png)
+
+Here you’ll see a handful of JSON-RPC calls that Hardhat/Ethers made under the hood for us when we called the .deploy() function. Two important ones to call out here are [eth_sendRawTransaction](https://docs.alchemyapi.io/alchemy/documentation/alchemy-api-reference/json-rpc#eth_sendrawtransaction), which is the request to actually write our smart contract onto the Ropsten chain, and [eth_getTrans ...
```
Can we reference these methods locally?
https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_sendrawtransaction
https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_gettransactionbyhash

wackerow reviewed (changes requested):
Nice, coming along @smudgil, getting closer. I uploaded the image assets for you to this folder. Left another round of comments, just ping me when ready =)

smudgil commented:
Hey Paul! Just submitted a bunch of edits- ready for some more feedback! Thanks!
[https://my-email-signature.link/signature.gif?u=1100946&e=149217439&v=15e1db3c131b4caf0fb7109f2ddb573fd9d2efdde512ca42932d69fd10b6beeb]
________________________________
From: Paul Wackerow ***@***.***>
Sent: Thursday, April 22, 2021 5:13 PM
To: ethereum/ethereum-org-website ***@***.***>
Cc: Sumi Pal Mudgil ***@***.***>; Mention ***@***.***>
Subject: Re: [ethereum/ethereum-org-website] Guide on how to code an NFT smart contract (#2652)


@wackerow requested changes on this pull request.

Nice, coming along @smudgil<https://github.com/smudgil>, getting closer. I uploaded the image assets for you to this folder. Left another round of comments, just ping me when ready =)

________________________________

In src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md<https://github.com/ethereum/ethereum-org-website/pull/2652#discussion_r618719405>:

> @@ -0,0 +1,341 @@

+---

+title: How to Write & Deploy an NFT (Part 1/3 of NFT Tutorial Series)

+description: This tutorial is Part 1 of a series on NFTs that will take you step by step on how to write and deploy a Non Fungible Token (ERC-721 token) smart contract using Ethereum and Inter Planetary File System (IPFS). In Part II, we will describe how to mint an NFT using our smart contract, and in Part III, we will explain how to view your freshly minted NFT on Metamask!

+author: "Sumi Mudgil"

+tags: ["NFTs", "ERC-721", "non fungible tokens", "Alchemy", "Solidity", "smart contracts"]



maybe pull the "non fungible tokens" entry from the list... ERC-721 and NFTs should do

________________________________

In src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md<https://github.com/ethereum/ethereum-org-website/pull/2652#discussion_r618721717>:

> +

+And of course, if you have questions at any point, don’t hesitate to reach out in the [Alchemy Discord](https://discord.gg/gWuC7zB)!

+

+## Step 1: Connect to the Ethereum network {#connect-to-ethereum}

+

+There a bunch of ways to make requests to the Ethereum blockchain, but to make things easiest, we’ll use a free account on [Alchemy](https://dashboard.alchemyapi.io/signup?referral=affiliate:3bda2c8a-2393-4923-a7d9-f98c7c2f4520), a blockchain developer platform and API that allows us to communicate with the Ethereum chain without having to run our own nodes.

+

+In this tutorial, we’ll also take advantage of Alchemy’s developer tools for monitoring and analytics to understand what’s going on under the hood in our smart contract deployment. If you don’t already have an Alchemy account, you can sign up for free [here](https://dashboard.alchemyapi.io/signup).

+

+## Step 2: Create your app (and API key) {#make-api-key}

+

+Once you’ve created an Alchemy account, you can generate an API key by creating an app. This will allow us to make requests to the Ropsten test network. Check out [this guide](https://docs.alchemyapi.io/guides/choosing-a-network) if you’re curious to learn more about test networks.

+

+1. Navigate to the “Create App” page in your Alchemy Dashboard by hovering over “Apps” in the nav bar and clicking “Create App”

+

+![Create your app](https://cdn-images-1.medium.com/max/5200/0*XQGWtaqrAdJNm5xq.png)



So I moved this markdown file into it's own subfolder, and renamed to /how-to-write-and-deploy-an-nft/index.md, and I downloaded the image assets your reference. I didn't otherwise change the markdown, if you wouldn't mind updating these images refs, (ie, here would be ![Create your app](./create-your-app.png)... just peak at the images I uploaded to double check the filenames.)

________________________________

In src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md<https://github.com/ethereum/ethereum-org-website/pull/2652#discussion_r618722841>:

> @@ -0,0 +1,341 @@

+---

+title: How to Write & Deploy an NFT (Part 1/3 of NFT Tutorial Series)

+description: This tutorial is Part 1 of a series on NFTs that will take you step by step on how to write and deploy a Non Fungible Token (ERC-721 token) smart contract using Ethereum and Inter Planetary File System (IPFS). In Part II, we will describe how to mint an NFT using our smart contract, and in Part III, we will explain how to view your freshly minted NFT on Metamask!

+author: "Sumi Mudgil"

+tags: ["NFTs", "ERC-721", "non fungible tokens", "Alchemy", "Solidity", "smart contracts"]

+skill: beginner

+lang: en

+sidebar: false

+published: 2020-04-22

+---

+With NFTs bringing blockchain into the public eye, now is an excellent opportunity to understand the hype yourself by publishing your own NFT (ERC-721 Token) on the Ethereum blockchain!

+

+Alchemy is extremely proud to be powering the biggest names in the NFT space, including Makersplace (recently set a record digital artwork sale at Christie’s for $69 Million), Dapper Labs (creators of NBA Top Shot & Crypto Kitties), OpenSea (the world’s largest NFT marketplace), Zora, Super Rare, NFTfi, Foundation, Enjin, Origin Protocol, Immutable, and more.

+

+In this tutorial, we will walk through creating and deploying an ERC-721 smart contract on the Ropsten test network using [Metamask](https://metamask.io/), [Solidity](https://docs.soliditylang.org/en/v0.8.0/), [Hardhat](https://hardhat.org/), [Pinata ](https://pinata.cloud/)and [Alchemy](https://dashboard.alchemyapi.io/signup?referral=affiliate:aab0c546-cc29-4180-8d31-fbce31367277) (don’t fret if you don’t understand what any of this means yet — we will explain it!).



The space after Pinata should not be within the link:
should be [Pinata](https://pinata.cloud/) and

________________________________

In src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md<https://github.com/ethereum/ethereum-org-website/pull/2652#discussion_r618736489>:

> +

+    //Contract based on [https://docs.openzeppelin.com/contracts/3.x/erc721](https://docs.openzeppelin.com/contracts/3.x/erc721)

+    // SPDX-License-Identifier: MIT

+    pragma solidity ^0.7.3;

+

+    import ***@***.***/contracts/token/ERC721/ERC721.sol";

+    import ***@***.***/contracts/utils/Counters.sol";

+    import ***@***.***/contracts/access/Ownable.sol";

+

+

+    contract MyNFT is ERC721, Ownable {

+        using Counters for Counters.Counter;

+        Counters.Counter private _tokenIds;

+

+        constructor() public ERC721("MyNFT", "NFT") {}

+

+        function mintNFT(address recipient, string memory tokenURI)

+            public onlyOwner

+            returns (uint256)

+        {

+            _tokenIds.increment();

+

+            uint256 newItemId = _tokenIds.current();

+            _mint(recipient, newItemId);

+            _setTokenURI(newItemId, tokenURI);

+

+            return newItemId;

+        }

+    }

+



So our markdown parser can sometimes get the indentation syntax correct, but it's safer to wrap these code blocks in triple-backticks. You can also add the language that way to help with syntax coloring.

This is how this section renders now:
[image]<https://user-images.githubusercontent.com/54227730/115784032-733e6400-a372-11eb-97d5-ca50a3071a0e.png>
If you wrap this block (and others) in triple-backticks (on their own lines) and trail the opening set with the language name, like this...
[image]<https://user-images.githubusercontent.com/54227730/115784411-eba52500-a372-11eb-8087-0649eff01c2e.png>
...then it will render like this:
[image]<https://user-images.githubusercontent.com/54227730/115784464-fe1f5e80-a372-11eb-8e1f-2ad92a8b898e.png>

Lets make sure all the code blocks are wrapped this way to minimize issues and keep it consistent =) thanks!

________________________________

In src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md<https://github.com/ethereum/ethereum-org-website/pull/2652#discussion_r618737592>:

> +

+So, what does this code *do* exactly? Let’s break it down, line-by-line.

+

+At the top of our smart contract, we import three [OpenZepplin](https://openzeppelin.com/) smart contract classes:

+

+* @openzeppelin/contracts/token/ERC721/ERC721.solcontains the implementation of the ERC-721 standard, which our NFT smart contract will inherit. (To be a valid NFT, your smart contract must implement all the methods of the ERC-721 standard.) To learn more about the inherited ERC-721 functions, check out the interface definition [here](https://eips.ethereum.org/EIPS/eip-721).

+

+* @openzeppelin/contracts/utils/Counters.solprovides counters that can only be incremented or decremented by one. Our smart contract uses a counter to keep track of the total number of NFTs minted and set the unique ID our new NFT. (Each NFT minted using a smart contract must be assigned a unique ID—here our unique ID is just determined by the total number of NFTs in existance. For example, the first NFT we mint with our smart contract has an ID of "1," our second NFT has an ID of "2," etc.)

+

+* @openzeppelin/contracts/access/Ownable.sol sets up [access control](https://docs.openzeppelin.com/contracts/3.x/access-control) on our smart contract, so only the owner of the smart contract (you) can mint NFTs. (Note, including access control is entirely a preference. If you'd like anyone to be able to mint an NFT using your smart contract, remove the word Ownable on line 10 and onlyOwner on line 17.)

+

+After our import statements, we have our custom NFT smart contract, which is surprisingly short — it only contains a counter, a constructor, and single function! This is thanks to our inherited OpenZepplin contracts, which implement most of the methods we need to create an NFT, such as ownerOf which returns the owner of the NFT, and transferFrom,which transfers ownership of the NFT from one account to another.

+

+In our ERC-721 constructor, you’ll notice we pass 2 strings, “MyNFT” and “NFT.” The first variable is the smart contract’s name, and the second is its symbol. You can name each of these variables whatever you wish!

+

+Finally, we have our function mintNFT(address recipient, string memory tokenURI) that allows us to mint an NFT! You'll notice it this function takes in two variables:



I'll leave this to your judgement, but I'm always a fan of inline code being styled as such by wrapping in single backticks inline.
[image]<https://user-images.githubusercontent.com/54227730/115784929-9e758300-a373-11eb-9185-db2418f0daf4.png>

________________________________

In src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md<https://github.com/ethereum/ethereum-org-website/pull/2652#discussion_r618738241>:

> +

+At the top of our smart contract, we import three [OpenZepplin](https://openzeppelin.com/) smart contract classes:

+

+* @openzeppelin/contracts/token/ERC721/ERC721.solcontains the implementation of the ERC-721 standard, which our NFT smart contract will inherit. (To be a valid NFT, your smart contract must implement all the methods of the ERC-721 standard.) To learn more about the inherited ERC-721 functions, check out the interface definition [here](https://eips.ethereum.org/EIPS/eip-721).

+

+* @openzeppelin/contracts/utils/Counters.solprovides counters that can only be incremented or decremented by one. Our smart contract uses a counter to keep track of the total number of NFTs minted and set the unique ID our new NFT. (Each NFT minted using a smart contract must be assigned a unique ID—here our unique ID is just determined by the total number of NFTs in existance. For example, the first NFT we mint with our smart contract has an ID of "1," our second NFT has an ID of "2," etc.)

+

+* @openzeppelin/contracts/access/Ownable.sol sets up [access control](https://docs.openzeppelin.com/contracts/3.x/access-control) on our smart contract, so only the owner of the smart contract (you) can mint NFTs. (Note, including access control is entirely a preference. If you'd like anyone to be able to mint an NFT using your smart contract, remove the word Ownable on line 10 and onlyOwner on line 17.)

+

+After our import statements, we have our custom NFT smart contract, which is surprisingly short — it only contains a counter, a constructor, and single function! This is thanks to our inherited OpenZepplin contracts, which implement most of the methods we need to create an NFT, such as ownerOf which returns the owner of the NFT, and transferFrom,which transfers ownership of the NFT from one account to another.

+

+In our ERC-721 constructor, you’ll notice we pass 2 strings, “MyNFT” and “NFT.” The first variable is the smart contract’s name, and the second is its symbol. You can name each of these variables whatever you wish!

+

+Finally, we have our function mintNFT(address recipient, string memory tokenURI) that allows us to mint an NFT! You'll notice it this function takes in two variables:

+

+* address recipient specifies the address that will receive your freshly minted NFT



Same here for example, as this looks like it should be capitalized at the beginning until you realize that address is referring to a variable, which would've been easier to spot with code styling

________________________________

In src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md<https://github.com/ethereum/ethereum-org-website/pull/2652#discussion_r618738898>:

> +

+* @openzeppelin/contracts/utils/Counters.solprovides counters that can only be incremented or decremented by one. Our smart contract uses a counter to keep track of the total number of NFTs minted and set the unique ID our new NFT. (Each NFT minted using a smart contract must be assigned a unique ID—here our unique ID is just determined by the total number of NFTs in existance. For example, the first NFT we mint with our smart contract has an ID of "1," our second NFT has an ID of "2," etc.)

+

+* @openzeppelin/contracts/access/Ownable.sol sets up [access control](https://docs.openzeppelin.com/contracts/3.x/access-control) on our smart contract, so only the owner of the smart contract (you) can mint NFTs. (Note, including access control is entirely a preference. If you'd like anyone to be able to mint an NFT using your smart contract, remove the word Ownable on line 10 and onlyOwner on line 17.)

+

+After our import statements, we have our custom NFT smart contract, which is surprisingly short — it only contains a counter, a constructor, and single function! This is thanks to our inherited OpenZepplin contracts, which implement most of the methods we need to create an NFT, such as ownerOf which returns the owner of the NFT, and transferFrom,which transfers ownership of the NFT from one account to another.

+

+In our ERC-721 constructor, you’ll notice we pass 2 strings, “MyNFT” and “NFT.” The first variable is the smart contract’s name, and the second is its symbol. You can name each of these variables whatever you wish!

+

+Finally, we have our function mintNFT(address recipient, string memory tokenURI) that allows us to mint an NFT! You'll notice it this function takes in two variables:

+

+* address recipient specifies the address that will receive your freshly minted NFT

+

+* string memory tokenURI is a string that should resolve to a JSON document that describes the NFT's metadata. An NFT's metadata is really what brings it to life, allowing it to have configurable properties, such as a name, description, image, and other attributes. In part 2 of this tutorial, we will describe how to configure this metadata.

+

+mintNFT calls some methods from the inherited ERC-721 library, and ultimately returns a number that represents the ID of the freshly minted NFT.



...and here. In particular, it's starting the sentence, so it should be capitalized, or styled as code to clarify.

________________________________

In src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md<https://github.com/ethereum/ethereum-org-website/pull/2652#discussion_r618740300>:

> +

+To make sure everything is working so far, let’s compile our contract. The compile task is one of the built-in hardhat tasks.

+

+From the command line run:

+

+    npx hardhat compile

+

+You might get a warning about SPDX license identifier not provided in source file , but no need to worry about that — hopefully everything else looks good! If not, you can always message in the [Alchemy discord](https://discord.gg/u72VCg3).

+

+## Step 15: Write our deploy script {#write-deploy}

+

+Now that our contract is written and our configuration file is good to go, it’s time to write our contract deploy script.

+

+Navigate to the `scripts/` folder and create a new file called `deploy.js` , adding the following contents to it:

+

+```



Like I mentioned before, you could also add js immediately following the opening set of triple backticks and then the syntax will be highlighted as JavaScript.

________________________________

In src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md<https://github.com/ethereum/ethereum-org-website/pull/2652#discussion_r618740779>:

> +

+## Step 14: Compile our contract {#compile-contract}

+

+To make sure everything is working so far, let’s compile our contract. The compile task is one of the built-in hardhat tasks.

+

+From the command line run:

+

+    npx hardhat compile

+

+You might get a warning about SPDX license identifier not provided in source file , but no need to worry about that — hopefully everything else looks good! If not, you can always message in the [Alchemy discord](https://discord.gg/u72VCg3).

+

+## Step 15: Write our deploy script {#write-deploy}

+

+Now that our contract is written and our configuration file is good to go, it’s time to write our contract deploy script.

+

+Navigate to the `scripts/` folder and create a new file called `deploy.js` , adding the following contents to it:



remove extra space after deploy.js

Navigate to the `scripts/` folder and create a new file called `deploy.js`, adding the following contents to it:



________________________________

In src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md<https://github.com/ethereum/ethereum-org-website/pull/2652#discussion_r618741683>:

> +

+If we go to the [Ropsten etherscan](https://ropsten.etherscan.io/) and search for our contract address we should able to see that it has been deployed successfully. The transaction will look something like this:

+

+![View your transaction address on Etherscan](https://cdn-images-1.medium.com/max/5200/0*qlORqkJ0KBb39Y1m.png)

+

+The From address should match your Metamask account address and the To address will say “Contract Creation.” If we click into the transaction, we’ll see our contract address in the To field:

+

+![View your contract address on Etherscan](https://cdn-images-1.medium.com/max/5200/0*yu-kGK524F12jFvZ.png)

+

+Yasssss! You just deployed your NFT smart contract to the Ethereum chain!

+

+To understand what’s going on under the hood, let’s navigate to the Explorer tab in our [Alchemy dashboard ](https://dashboard.alchemyapi.io/explorer). If you have multiple Alchemy apps make sure to filter by app and select “MyNFT”.

+

+![View calls made “under the hood” with Alchemy’s Explorer Dashboard](https://cdn-images-1.medium.com/max/5200/0*9vAulgWSnJJ_Xf4u.png)

+

+Here you’ll see a handful of JSON-RPC calls that Hardhat/Ethers made under the hood for us when we called the .deploy() function. Two important ones to call out here are [eth_sendRawTransaction](https://docs.alchemyapi.io/alchemy/documentation/alchemy-api-reference/json-rpc#eth_sendrawtransaction), which is the request to actually write our smart contract onto the Ropsten chain, and [eth_getTransactionByHash](https://docs.alchemyapi.io/alchemy/documentation/alchemy-api-reference/json-rpc#eth_gettransactionbyhash) which is a request to read information about our transaction given the hash (a typical pattern when sending transactions). To learn more about sending transactions, check out this tutorial on [sending transactions using Web3](https://docs.alchemyapi.io/alchemy/tutorials/sending-transactions-using-web3-and-alchemy).



Can we reference these methods locally?
https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_sendrawtransaction
https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_gettransactionbyhash

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ethereum/ethereum-org-website/pull/2652#pullrequestreview-642641889>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AJDNA67OGEP7U2G2J5EZ2J3TKCGPBANCNFSM4ZIXHASQ>.

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md:
```diff
@@ -1,18 +1,18 @@
 ---
 title: How to Write & Deploy an NFT (Part 1/3 of NFT Tutorial Series)
-description: This tutorial is Part 1 of a series on NFTs that will take you step by step on how to write and deploy a Non Fungible Token (ERC-721 token) smart contract using Ethereum and Inter Planetary File System (IPFS). In Part II, we will describe how to mint an NFT using our smart contract, and in Part III, we will explain how to view your freshly minted NFT on Metamask!
+description: This tutorial is Part 1 of a series on NFTs that will take you step by step on how to write and deploy a Non Fungible Token (ERC-721 token) smart contract using Ethereum and Inter Planetary File System (IPFS). In [Part II, we will describe how to mint an NFT using our smart contract](/developers/tutorials/how-to-mint-an-nft/), and in [Part III, we will explain how to view your freshl ...
```
See my comment on Part 3... would trim this down to one sentence without links

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md:
```diff
@@ -1,18 +1,18 @@
 Alchemy is extremely proud to be powering the biggest names in the NFT space, including Makersplace (recently set a record digital artwork sale at Christie’s for $69 Million), Dapper Labs (creators of NBA Top Shot & Crypto Kitties), OpenSea (the world’s largest NFT marketplace), Zora, Super Rare, NFTfi, Foundation, Enjin, Origin Protocol, Immutable, and more.
 
-In this tutorial, we will walk through creating and deploying an ERC-721 smart contract on the Ropsten test network using [Metamask](https://metamask.io/), [Solidity](https://docs.soliditylang.org/en/v0.8.0/), [Hardhat](https://hardhat.org/), [Pinata ](https://pinata.cloud/)and [Alchemy](https://dashboard.alchemyapi.io/signup?referral=affiliate:aab0c546-cc29-4180-8d31-fbce31367277) (don’t fret if ...
+In this tutorial, we will walk through creating and deploying an ERC-721 smart contract on the Ropsten test network using [Metamask](https://metamask.io/), [Solidity](https://docs.soliditylang.org/en/v0.8.0/), [Hardhat](https://hardhat.org/), [Pinata](https://pinata.cloud/)and [Alchemy](https://dashboard.alchemyapi.io/signup?referral=affiliate:aab0c546-cc29-4180-8d31-fbce31367277) (don’t fret if  ...
```
(Noted this on Part 3, but MetaMask uses two capitals, missed this before, sorry)

wackerow replied:
and you removed the space that was in the wrong place, but we need it after the link. currently renders:
[Pinata](https://pinata.cloud/)and

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md:
```diff
@@ -30,11 +30,11 @@ Once you’ve created an Alchemy account, you can generate an API key by creatin
 2. Name your app (we chose “My First NFT!”), offer a short description, select “Staging” for the Environment (used for your app bookkeeping), and choose “Ropsten” for your network.
 
-![Configure and publish your app](https://cdn-images-1.medium.com/max/5200/0*AFYHVgRTeBsEfTPR.png)
+![Configure and publish your app](./configure-and-publish-your-app.png )
```
stray space after png

wackerow commented on src/content/developers/tutorials/how-to-write-and-deploy-an-nft/index.md:
```diff
@@ -149,10 +149,11 @@ Now time that our environment is set up, onto more exciting stuff: *writing our
 Open up the my-nft project in your favorite editor (we like [VSCode](https://code.visualstudio.com/)). Smart contracts are written in a language called Solidity which is what we will use to write our MyNFT.sol smart contract.‌
 
-1. Navigate to the “contracts” folder and create a new file called MyNFT.sol
+1. Navigate to the `“contracts”` folder and create a new file called MyNFT.sol
```
Just the back-ticks will do, no need for " "... but while on this line, could extend that to 
```
1. Navigate to the `contracts` folder and create a new file called `MyNFT.sol`
```
1. Navigate to the `contracts` folder and create a new file called `MyNFT.sol`

minimalsm commented:
Hey @smudgil! 

Just read through this for the first time and I think it's great 🥰! Anything I can do to help you get this over the line and live on the site?

minimalsm commented:
Hey @smudgil 👋 Just to keep you in the loop I've made some changes across the three tutorials to get them over the line and I'm merging them now. Thanks a bunch for all your work on these tutorials and being cooperative on the editing too 😀

minimalsm commented:
@all-contributors please add @smudgil for content
